kernel/src/snapshot.rs

```diff
@@ -239,6 +366,211 @@ mod tests {
         assert_eq!(snapshot.schema(), expected);
     }
 
+    // interesting cases for testing Snapshot::new_from:
+    // 1. new version < existing version
+    // 2. new version == existing version
+    // 3. new version > existing version AND
+    //   a. log segment hasn't changed
+    //   b. log segment for old..=new version has a checkpoint (with new protocol/metadata)
+    //   b. log segment for old..=new version has no checkpoint
+    //     i. commits have (new protocol, new metadata)
+    //     ii. commits have (new protocol, no metadata)
+    //     iii. commits have (no protocol, new metadata)
+    //     iv. commits have (no protocol, no metadata)
```

Comment:
I've tried to roughly outline the tests here, I think we have a bit of work to do to make testing easier but instead of invest a lot of time here i'm proposing we ensure we have coverage but then take a follow-up to clean up testing (share code, etc.) just so we can introduce this API and not have it linger for too much longer

Reply:
https://github.com/delta-io/delta-kernel-rs/issues/811